twister: refine the twister test plan generation

smrtos opened the pull request:
(This patch is a refined version of https://github.com/zephyrproject-rtos/zephyr/pull/46052 to cater for recent twister refactoring.)

Due to the diverse coding styles and lack of preprocessing when
scanning for test cases, there were many unintended combinations
of yaml test scenarios and C test functions with the regex-based
test case discovery, which caused an inaccurate test plan and test
result stats.

As the new ztest fx is used, the test cases of a test instance can
be accurately determined via zephyr.symbols file.

Because the zephyr.symbols file is not available until after build,
test cases determination is placed right after the build stage of
the pipeline and before the runtime test collection.

For those test instances that don't go through the build stage,
such as those with "Skip filtered" (statically filtered) reason
before entering the pipeline, they are not affected.

This patch also adjust the stats logic because the zephyr.symbols
file is generated after build. So ExecutionCounter update is split
and some must be postponed until test execution pipeline is completed.

Some concepts:

A test instance = a yaml scenario + a platform
"Test instance" and "test config" are synonyms in twister, unfortunately
excessive IMHO...

A test instance can be filtered or skipped in 3 ways.
Let's define "runtime" as "after entering the execution pipeline".

1) statically filtered (before runtime)
   Such test instance is filtered by analyzing the yaml and never
   enters the execution pipeline.
2) cmake filtered (runtime)
   Such test instance enters pipeline and is filtered at cmake stage.
3) build skipped (also runtime)
   Such test instance enters pipeline and is skipped at build stage.

All the test instances that enter the execution pipeline will go
through the report stage, where ExecutionCounter is further updated.

The meaning of the fields of ExecutionCounter are:

 .skipped_configs = static filtered + cmake filtered + build skipped
 .skipped_runtime = cmake filtered + build skipped
 .skipped_filter = static filtered
 .done = instances that enter the execution pipeline
 .passed = instances that actually executed and passed

Definition of the overall stats:
total_complete = .done + .skipped_filter
total = yaml test scenario * applicable platforms
complete percentage = total_complete / total
pass rate = passed / (total - skipped_configs)


Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos committed:
twister: refine the twister test plan generation

Due to the diverse coding styles and lack of preprocessing when
scanning for test cases, there were many unintended combinations
of yaml test scenarios and C test functions with the regex-based
test case discovery, which caused an inaccurate test plan and test
result stats.

As the new ztest fx is used, the test cases of a test instance can
be accurately determined via zephyr.symbols file.

Because the zephyr.symbols file is not available until after build,
test cases determination is placed right after the build stage of
the pipeline and before the runtime test collection.

For those test instances that don't go through the build stage,
such as those with "Skip filtered" (statically filtered) reason
before entering the pipeline, they are not affected.

This patch also adjust the stats logic because the zephyr.symbols
file is generated after build. So ExecutionCounter update is split
and some must be postponed until test execution pipeline is completed.

Some concepts:

A test instance = a yaml scenario + a platform
"Test instance" and "test config" are synonyms in twister, unfortunately
excessive IMHO...

A test instance can be filtered or skipped in 3 ways.
Let's define "runtime" as "after entering the execution pipeline".

1) statically filtered (before runtime)
   Such test instance is filtered by analyzing the yaml and never
   enters the execution pipeline.
2) cmake filtered (runtime)
   Such test instance enters pipeline and is filtered at cmake stage.
3) build skipped (also runtime)
   Such test instance enters pipeline and is skipped at build stage.

All the test instances that enter the execution pipeline will go
through the report stage, where ExecutionCounter is further updated.

The meaning of the fields of ExecutionCounter are:

 .skipped_configs = static filtered + cmake filtered + build skipped
 .skipped_runtime = cmake filtered + build skipped
 .skipped_filter = static filtered
 .done = instances that enter the execution pipeline
 .passed = instances that actually executed and passed

Definition of the overall stats:
total_complete = .done + .skipped_filter
total = yaml test scenario * applicable platforms
complete percentage = total_complete / total
pass rate = passed / (total - skipped_configs)

Signed-off-by: Ming Shao <ming.shao@intel.com>

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -511,6 +517,47 @@ def process(self, pipeline, done, message, lock, results):
+        if os.path.isfile(symbol_file):
+            logger.debug(f"zephyr.symbols located: {symbol_file}")
+        else:
+            # no zephyr.lst file, cannot do symbol-based amending
```
we do not use .lst file anymore

smrtos replied:
will fix that.

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -511,6 +517,47 @@ def process(self, pipeline, done, message, lock, results):
             else:
                 self.cleanup_artifacts()
 
+    def amend_test_instance(self, results):
```
why amend? The goal is to fuilly rely on what comes out of the symbols file as the list of testcases in the suite. This should just replaces whatever is in there and should be seen as the true source.

At this stage we should also not skip anything and should make judgment about unintended tests, this should purly serve as a replacement for regex method, however, only for ztest with new API for now.

We also want to capture the testsuite names from the symbols file.

smrtos replied:
Please see my next reply.

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -511,6 +517,47 @@ def process(self, pipeline, done, message, lock, results):
+        with open(symbol_file, 'r') as fp:
+            symbols = fp.read()
+            logger.debug(f"test instance {self.instance.name} has {len(self.instance.testcases)} cases.")
+            for testcase in self.instance.testcases:
```
That is where I am getting confused, you are checking against what was parsed earlier using regex, but the intent is to replace that and just use the input from the symbols file

smrtos replied:
I was planning to solely rely on the zephyr.symbols file. 

But it cannot handle below scenario:

https://github.com/zephyrproject-rtos/zephyr/blob/b2f80b0e16363021bc1fd65884b11830fa869870/tests/lib/cmsis_dsp/matrix/src/binary_q7.c#L119-L126

The `test_matrix_binary_q7()` and `test_op2_arm_mat_mult_q7()` both have `test_` prefix but only the latter is a valid unit test. I cannot tell them apart just with the `zephyr.symbols` file.  Even `test_main()` has a `test_` prefix.

Though the symbol file does show some subtle difference between them as shown below. The `T` means global text section and `t` means local text section.
```
00000ca8 T test_matrix_binary_q7
00000a0c t test_op2_arm_mat_mult_q7
```
But that's not always true because in our code, a unit test function can be declared with or without `static` keyword.
For example below one is a unit test without `static`:

https://github.com/zephyrproject-rtos/zephyr/blob/b2f80b0e16363021bc1fd65884b11830fa869870/tests/crypto/tinycrypt/src/aes.c#L71

So, we have to keep the old regex-based method for the sake of old-style tests.

If we add another test collection method for the new ztest fx **right now**, that means we may need to tell which method to use when scanning for tests.  We may make a choice by checking for `CONFIG_NEW_ZTEST` in the final `.config`.

But the regex-based collection happens **before** build, while symbol-based collection happens **after** build. Currently the regex-based collection **can** collect for both styles. While the symbol-based collection can only cover the new style. When the time comes to carry out symbol-based collection, all tests have **already** been collected, though with some falseness. So I choose to amend rather than re-collect the new-style tests.

Actually I consider this PR as just **phase 1**, which aims to remove false test suite/test function combinations such as in cmsis tests.

In future, when we fully migrate to the new ztest fx. A **phase 2** PR can be done to solely rely on the `zephyr.symbols`. By then, we will be able to readily use the `z_ztest_unit_test` prefix to find unit tests and `z_ztest_test_node` to find test suite.

nashif replied:
> So, we have to keep the old regex-based method for the sake of old-style tests.

I am not suggesting it should be removed now.

The tests you mentioned above were not converted to new ztest API yet and those still need to be parsed using regex, you should not try and get the tests from binaries of tests that were not converted.

> If we add another test collection method for the new ztest fx right now, that means we may need to tell which method to use when scanning for tests.

In that case you will be using the symbols file itself, if you are using some regex to collect the data (using new ztest api) then you will get no matches with old tests and can skip in that case.

> So I choose to amend rather than re-collect the new-style tests.

I would just override 

> Actually I consider this PR as just phase 1, which aims to remove false test suite/test function combinations such as in cmsis tests.

Yes, this is phase 1, I agree.

smrtos replied:
> In that case you will be using the symbols file itself, if you are using some regex to collect the data (using new ztest api) then you will get no matches with old tests and can skip in that case.

I think my current way to utilize the `zephyr.symbols` for amending can benefit both old and new styles tests.
Maybe this is an optimum result for phase 1?

> I would just override

I feel I am kind of overriding the regex-parsing results now. Though I call it amend. 
Do you mean besides amending the invalid ones from regex parsing, I should also **re-add** the valid ones found with symbols, in spite of that they have already been found through regex parsing?

Is that really necessary since I saw the regex-parsing logic which caters for the new ztest:

https://github.com/zephyrproject-rtos/zephyr/blob/b2f80b0e16363021bc1fd65884b11830fa869870/scripts/pylib/twister/twisterlib/testsuite.py#L76-L78

smrtos replied:
@nashif  Just added logic to override the old regex-based test case collection result. 
And it should also work when no regex-based test collection exists. Will test more on that.

smrtos replied:
@nashif  Updated the commit message to reflect what's going on.

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -511,6 +517,47 @@ def process(self, pipeline, done, message, lock, results):
+                # for the new-style ztest function
+                test_func_name2_regex = re.compile("z_ztest_unit_test_.*" + identifier.split('.')[-1])
+
+                # If the .lst file exists and the test case function not in it,
```
.lst -> symbols file

smrtos replied:
Will fix the comment.

nashif commented on scripts/pylib/twister/twisterlib/testsuite.py:
```diff
@@ -385,6 +386,7 @@ def load(self, data):
     def add_subcases(self, data, parsed_subcases, suite_names):
         testcases = data.get("testcases", [])
         if testcases:
+            self.has_explicit_testcases_in_yaml = True
```
that is a veeeeeeeeeery long variable :)

smrtos replied:
Yeah, I will figure out a new one.

smrtos commented:
I forgot to change one variable name in the `runner.py` in last force push. But it still passed all the CI tests. Amazing...

smrtos commented:
Made chagnes and updated the commit message to reflect what's going on.

smrtos commented:
Compared with the old test plan with below command:

> twister -p native_posix -T tests/lib/cmsis_dsp/matrix/ -vv

With the **old** test plan generation:
![image](https://user-images.githubusercontent.com/54129187/191641083-24f5f43a-319a-4ebc-9a27-251db35d470b.png)


With the **new** test plan generation:
![image](https://user-images.githubusercontent.com/54129187/191641154-f8cea09f-e61d-4ab6-80e6-ec3eb198aafb.png)

We can see a lot of false test combination are identified. They are like below:

```
{
    "identifier":"libraries.cmsis_dsp.matrix.unary_q7.arm_mat_solve_upper_triangular_f32",
    "execution_time":"0.00",
    "status":"skipped",
    "reason":"Test case not belong to the test scenario"
},
```

And also, with this PR, the `No status` messages like below can be eliminated.

![image](https://user-images.githubusercontent.com/54129187/191641454-32c4b233-9214-412b-84f9-6ff1a253129d.png)

yperess commented:
@asemjonovs can you take a look too?

smrtos commented:
Further tested with below commands.
The numbers of the same color match.

> twister -p qemu_cortex_m3 -T tests/lib/cmsis_dsp/matrix/

![image](https://user-images.githubusercontent.com/54129187/191673236-8312ea55-a076-4980-9a8a-5089454ee3e8.png)


> twister -p qemu_cortex_m3 -T tests/lib/cmsis_dsp/matrix/ **-vv**

![image](https://user-images.githubusercontent.com/54129187/191673328-fd8036b5-05d0-4688-bc1e-32ce1f570e70.png)

asemjonovs commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -511,6 +546,71 @@ def process(self, pipeline, done, message, lock, results):
+            for m in matches:
+                # new_ztest_suite = m[0] # not used for now
+                test_func_name = m[1].replace("test_", "")
+                testcase_id = f"{yaml_testsuite_name}.{test_func_name}"
```
Can we include `new_ztest_suite` in the testcase_id?  I feel this info is lost in the organization of test report. So something like 
```
testcase_id = f"{yaml_test_config}.{ztest_suite}.{ztest_name}"
```

smrtos replied:
I did think about this before.  Line 605 is the relic of my thought. 

If I add the test case including the test suite, like `{yaml_testsuite_name}.{new_ztest_suite}.{test_func_name}`, when the harness sees a test case in the runtime serial log with the name of `{yaml_testsuite_name}.{test_func_name}`, there's no ztest suite info in the log line, so it will still consider it as a missing test case. And add it using below logic.

https://github.com/zephyrproject-rtos/zephyr/blob/8c7eba346af3119f5920a288af93f0fa5efd76eb/scripts/pylib/twister/twisterlib/harness.py#L14

https://github.com/zephyrproject-rtos/zephyr/blob/8c7eba346af3119f5920a288af93f0fa5efd76eb/scripts/pylib/twister/twisterlib/harness.py#L273-L275

https://github.com/zephyrproject-rtos/zephyr/blob/1bbe209a30fcdab037c8ac9b96c33ea2249665e0/scripts/pylib/twister/twisterlib/testinstance.py#L111-L119


This will cause incorrect stats.  And the newly added one will still be considered as `No status`.


For now I prefer to leave the `new_ztest_suite` out of testcase id. When all tests have been migrated to the new ztest framework, maybe we can leverage the test summary (https://github.com/zephyrproject-rtos/zephyr/pull/48499) to collect the result.  That'll be simpler and more accurate.

asemjonovs commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -511,6 +546,71 @@ def process(self, pipeline, done, message, lock, results):
+            logger.debug(f"zephyr.symbols NOT found: {symbol_file}")
+            return
+
+        yaml_testsuite_name = self.instance.testsuite.id
```
I see the yaml as more of a test config than suite, could we rename this?

smrtos replied:
Yes, yaml file is a test config file.
But unfortunately, it is represented with the `class TestSuite`:
https://github.com/zephyrproject-rtos/zephyr/blob/384ff96b255b03b92aee50d5d036cb78971950b8/scripts/pylib/twister/twisterlib/testsuite.py#L339

So I follow it to avoid more confusion in jargons.

smrtos commented:
BTW, I added below info about test instances.

```
twister -p native_posix -T tests/lib/cmsis_dsp/matrix/
```

> NFO    - Test instances brief:
> INFO    - (1) 12 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix (filtered)
> INFO    - (2) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_q7 (passed)
> INFO    - (3) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_q7.fpu (passed)
> INFO    - (4) 18 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_q15 (passed)
> INFO    - (5) 18 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_q15.fpu (passed)
> INFO    - (6) 18 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_q31 (passed)
> INFO    - (7) 18 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_q31.fpu (passed)
> INFO    - (8) 12 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_f16 (filtered)
> INFO    - (9) 12 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_f16.fpu (filtered)
> INFO    - (10) 18 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_f32 (passed)
> INFO    - (11) 18 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_f32.fpu (passed)
> INFO    - (12) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_f64 (passed)
> INFO    - (13) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.unary_f64.fpu (passed)
> INFO    - (14) 13 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_q7 (passed)
> INFO    - (15) 13 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_q7.fpu (passed)
> INFO    - (16) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_q15 (passed)
> INFO    - (17) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_q15.fpu (passed)
> INFO    - (18) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_q31 (passed)
> INFO    - (19) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_q31.fpu (passed)
> INFO    - (20) 12 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_f16 (filtered)
> INFO    - (21) 12 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_f16.fpu (filtered)
> INFO    - (22) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_f32 (passed)
> INFO    - (23) 14 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_f32.fpu (passed)
> INFO    - (24) 13 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_f64 (passed)
> INFO    - (25) 13 cases in: native_posix/tests/lib/cmsis_dsp/matrix/libraries.cmsis_dsp.matrix.binary_f64.fpu (passed)
>

nashif commented:
> BTW, I added below info about test instances.

Please remove this, this might look helpful running on one test and one platform, but it will flood the screen when the numbers grow. For all of this, we have the test output that can always be looked at.

smrtos commented on subsys/testsuite/ztest/include/zephyr/ztest_test_new.h:
```diff
@@ -336,7 +336,7 @@ void ztest_test_skip(void);
 	static void _##suite##_##fn##_wrapper(void *data);                                         \
 	static void suite##_##fn(                                                                  \
 		COND_CODE_1(use_fixture, (struct suite##_fixture *fixture), (void)));              \
-	static STRUCT_SECTION_ITERABLE(ztest_unit_test, z_ztest_unit_test_##suite##_##fn) = {      \
```
I will see how to split the commits more reasonably.

smrtos replied:
This change enables twister to tell the new ztest suite name and function name, which is necessary to compose `testcsae_id` for the test case based on zephyr.symbols.  Changing twister alone cannot make it. So I think it logically belongs to one commit for this single purpose.

smrtos commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -515,6 +550,72 @@ def process(self, pipeline, done, message, lock, results):
+            for m in matches:
+                # new_ztest_suite = m[0] # not used for now
+                test_func_name = m[1].replace("test_", "")
+                testcase_id = f"{yaml_testsuite_name}.{test_func_name}"
```
I will refine the logic.

smrtos replied:
> 1) check if we have matches in the symbols file for new ztest tests
> 2) if yes, remove all existing testcases and only use the newly parsed testcase names
> 3) if no, leave it as is.

I agree with 1) and 2). And 2) is a more elegant solution.

But for 3), if we leave the old ztest tests as is, I am afraid we will still miss the false combinations in there, which beats part of the purpose of this PR. So amending is still necessary in this transitional phase.

nashif replied:
> But for 3), if we leave the old ztest tests as is, I am afraid we will still miss the false combinations in there, which beats part of the purpose of this PR. So amending the old ztest tests is still necessary in this transitional phase.

what false combination? We are almost at 90% in the transition, in a few weeks we will switch to only new ztest API and the problematic tests have been all converted.  Lets not spend time on adding logic that will not be needed really soon and that would only have been useful with problematic testcases.

smrtos replied:
@nashif 

The old ztest are still being used for below tests:

> tests/bluetooth
> tests/drivers
> tests/net
> tests/ztest
> 

I run the all the `/tests` with below command:

```
twister --hardware-map ~/sources/sm_zephyr_tools/twister_hw_map/reel.map --device-testing -x=CONFIG_BOOT_DELAY=1500 -T tests/
```

And only found 2 false combinations:

```
{
    "identifier":"libraries.libc.picolibc.mem_alloc.no_mem_realloc",
    "execution_time":"0.00",
    "status":"skipped",
    "reason":"Test case doesn't belong to the test scenario"
},
{
    "identifier":"libraries.libc.picolibc.mem_alloc.no_mem_malloc",
    "execution_time":"0.00",
    "status":"skipped",
    "reason":"Test case doesn't belong to the test scenario"
},
```

So I removed the amending logic.
Should be OK.

smrtos commented:
> > BTW, I added below info about test instances.
> 
> Please remove this, this might look helpful running on one test and one platform, but it will flood the screen when the numbers grow. For all of this, we have the test output that can always be looked at.

Removed.

smrtos commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -27,22 +27,49 @@
 
 class ExecutionCounter(object):
     def __init__(self, total=0):
```
Do you mean the comments to the fields? I can move them to a separate commit.

smrtos commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -468,6 +495,14 @@ def process(self, pipeline, done, message, lock, results):
+                    # Collect and amend the test cases for the test instance
+                    # This should only happen when test author didn't explicitly list the
+                    # test cases in testcase.yaml.
+                    if not self.instance.testsuite.testcases_in_yaml:
```
Removed the `testcases_in_yaml` check.

smrtos commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -515,6 +550,72 @@ def process(self, pipeline, done, message, lock, results):
             else:
                 self.cleanup_artifacts()
 
+    # Both old and new ztest fx co-exist for now.
```
Fixed.

smrtos commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -650,7 +753,7 @@ def report_out(self, results):
                 Fore.RESET,
                 completed_perc,
                 Fore.YELLOW if results.skipped_configs > 0 else Fore.RESET,
-                results.skipped_filter + results.skipped_runtime,
```
This is not a bug. I just found a more concise and equivalent way to do it.

smrtos replied:
This is not a bug. I just found a more concise and equivalent way and can't help to replace it.

smrtos commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -817,19 +923,28 @@ def run(self):
             if retries == 0 or (self.results.failed == self.results.error and not self.options.retry_build_errors):
                 break
 
-    def update_counting(self):
```
The stats logic also needs to be adjusted because the zephyr.symbols is
generated after build. So ExecutionCounter update is split and some
must be postponed until test execution pipeline is completed.

smrtos committed:
twister: add comment about stats logic

Add comments and docstring to class  ExecutionCounter to explain
the twsiter stats logic.

Signed-off-by: Ming Shao <ming.shao@intel.com>

smrtos commented:
Removed the amending logic.
Update the commit msg.

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -27,29 +27,61 @@
-        logger.debug(f"Skipped Testsuites: {self.skipped_configs}")
-        logger.debug(f"Skipped Testsuites (runtime): {self.skipped_runtime}")
-        logger.debug(f"Skipped Testsuites (filter): {self.skipped_filter}")
+        logger.debug(f"Completed test instances: {self.done}")
```
instance is internal to twister, we should keep this as testsuite, because this is exactly what will show up in the test report, a testsuite.

smrtos replied:
Didn't realize that we want to encapsulate the concept of "test instance".
Will change it.

smrtos replied:
But the stats like `skipped_configs` are all calculated at the test instance level. Such as below:
https://github.com/zephyrproject-rtos/zephyr/blob/d43e4e9ca4e80055e9a0f0b095c856103ee0d11e/scripts/pylib/twister/twisterlib/runner.py#L602-L604

I can change back to the `testsuite` in the log message. But it's not really what the numbers mean.

nashif replied:
still, instance is historical and internal to twister. Please just keep it as is and stick to the original purpose of the PR.

smrtos replied:
@nashif 

Got it. Just re-pushed.

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -27,29 +27,61 @@
         logger.debug("--------------------------------")
-        logger.debug(f"Total Test suites: {self.total}")
-        logger.debug(f"Total Test cases: {self.cases}")
+        logger.debug(f"Total test suites: {self.total}") # actually test instances
```
you can't resist, can you?
